### PR TITLE
ztest: allow asserts anywhere

### PR DESCRIPTION
**Pull request by yperess**
Updates the ztest_test_fail() function to allow failures in setup.
When executed, a failed assert will fail every test in the suite owning
the setup function. This was verified by adding a suite which asserts
in the setup function and has a test that should pass. During
exeuction, ztest marks the test as failing. In order to keep the test
I also modified twister to include a new specification:
`expected_failed_tests`. When a test is marked as failed, twister will
ignore the failure and treat it as passing.

### CODE REVIEW

**Comment by yperess** [comment]
This is my first pass at allowing failed asserts in setup. It works, but I'm having a hard time adding a unit test for this feature. Unlike Java or other high level languages, we can't really annotate the test function as expected to fail. Since we don't run the test function we can't include a public ztest API telling it to expect the failure.

The update to twister is clean, but there's an issue, the final log `PROJECT EXECUTION FAILED` will prevail even if all tests are marked as passing.

One alternative I'm currently considering is adding a new macro to ztest:
```
ZTEST_EXPECT_FAIL(suite, test);
```

This will create a list of:
```
struct ztest_expected_failed_tests {
  const char *suite_name;
  const char *test_name;
};
```

With this, ztest will be able to actually set the state of the test as passing if failed, or failed if passing. I just wasn't a fan of having it in the linker like that, but I'm not seeing another good alternative. If anyone has any ideas I'll be glad to hear them, otherwise I'll update this PR to use the above macro instead of the twister changes and mark it as ready for review.

File: subsys/testsuite/ztest/src/ztest_new.c [hunk 1]
```diff
@@ -526,7 +577,12 @@ static int z_ztest_run_test_suite_ptr(struct ztest_suite_node *suite)
 	TC_SUITE_START(suite->name);
 	test_result = ZTEST_RESULT_PENDING;
 	phase = TEST_PHASE_SETUP;
-	if (suite->setup != NULL) {
+#ifndef KERNEL
+	if (setjmp(test_suite_fail)) {
+		test_result = ZTEST_RESULT_SUITE_FAIL;
+	}
+#endif
```
Comment (yperess) [inline subsys/testsuite/ztest/src/ztest_new.c:584]: @asemjonovs I think I need `#else` here to handle the signals safely and not kill the main thread.

File: subsys/testsuite/ztest/src/ztest_new.c [hunk 2]
```diff
@@ -348,7 +395,7 @@ static void test_finalize(void)
 
 void ztest_test_fail(void)
 {
-	test_result = ZTEST_RESULT_FAIL;
+	test_result = (phase == TEST_PHASE_SETUP) ? ZTEST_RESULT_SUITE_FAIL : ZTEST_RESULT_FAIL;
 	test_finalize();
```
Comment (asemjonovs) [inline subsys/testsuite/ztest/src/ztest_new.c:399]: This function is being invoked by the assert during setup.  `test_finalize` kills the current thread.  I think we need to conditionally call `test_finalize` when `phase != TEST_PHASE_SETUP`

Comment (yperess) [reply subsys/testsuite/ztest/src/ztest_new.c:399]: Thanks, I see the same below for skip.

**Comment by andyross** [comment]
Some drive by notes, with the obligatory reminder that this isn't the level of the stack where I'm best prepared to have ideas:

> With this, ztest will be able to actually set the state of the test as passing if failed, or failed if passing

Maybe not do this in C at all?  Have the rig in twister[1] understand that "this case must fail" via configuration in testcase.yaml, or have a much smaller feature in ztest emit some kind of "expect this suite/case to fail" message that can be interpreted by the wrapper to invert the sense of success, etc...

Ultimately, this is a feature in ztest that only exists so that ztest itself can be tested.  That's usually a sign that the test is at the wrong level of abstraction.

[1] Or better: a unified "evaluate ztest device output vs. test definition" library unified between west and twister.  It's frustrating that so much of our build/run logic ends up duplicated twice in two parallel (and very complicated) python tools.

**Comment by aaronemassey** [comment]
Just an FYI:

Seems like we have two separate conversations going on in this PR:

1) The actual changes made by the PR, where we can use zassert in setup.

2) Facilitating verification of failures.

We should probably move discussion of 2) to either an issue or its own PR.

**Comment by aaronemassey** [comment]
I didn't realize we were making two API changes in this PR.

1) Allowing zasserts anywhere
2) Annotating ZTESTs to expect failure or skip.

We probably should separate these as two different commits. With the understanding that one commits tests the prior.

File: doc/develop/test/ztest.rst [hunk 3]
```diff
@@ -110,6 +110,35 @@ This is achieved via fixtures in the following way::
...
+Some tests were made to be broken. In cases where the test is expected to fail or skip due to the
+nature of the code, it's possible to annotate the test as such. For example::
+
+  .. code-block:: C
+
+  #include <zephyr/ztest.h>
+
+  ZTEST_SUITE(my_suite, NULL, NULL, NULL, NULL, NULL);
+
+  ZTEST_EXPECT_FAIL(my_suite, test_fail)
```
Comment (aaronemassey) [inline doc/develop/test/ztest.rst:125]: Can we separate the two API changes as their own commits? It'll be a bit easier to review and understand which change is related to what API?

**Review by aaronemassey — CHANGES_REQUESTED** [review]

**Comment by yperess** [comment]
> Some drive by notes, with the obligatory reminder that this isn't the level of the stack where I'm best prepared to have ideas:
> 
> > With this, ztest will be able to actually set the state of the test as passing if failed, or failed if passing
> 
> Maybe not do this in C at all?  Have the rig in twister[1] understand that "this case must fail" via configuration in testcase.yaml, or have a much smaller feature in ztest emit some kind of "expect this suite/case to fail" message that can be interpreted by the wrapper to invert the sense of success, etc...
> 
> Ultimately, this is a feature in ztest that only exists so that ztest itself can be tested.  That's usually a sign that the test is at the wrong level of abstraction.
> 
> [1] Or better: a unified "evaluate ztest device output vs. test definition" library unified between west and twister.  It's frustrating that so much of our build/run logic ends up duplicated twice in two parallel (and very complicated) python tools.

I agree, and what you're describing was my first stab at this. I added a key to Twister's testcase.yaml (`expected_failed_tests`) and I was able to apply the same logic that I added here into twister. But... The very last statement from zmake tells twister if the whole binary passed/failed. As an example if you compiled some tests into your binary but never ran them, we have a hook to signal that the binary is a failure. When that happens you can get all passing tests but still get an error. There's no good override for that. In this example, if we fail a test and twister overrides that too a pass, the binary will still be considered as a fail and twister will have no way to tell if it should be overridden.

This approach is similar to using Java annotations so there's some precedent (though I agree that in twister would be nice)

**Comment by yperess** [comment]
> I didn't realize we were making two API changes in this PR.
> 
> 1) Allowing zasserts anywhere
> 2) Annotating ZTESTs to expect failure or skip.
> 
> We probably should separate these as two different commits. With the understanding that one commits tests the prior.

In what scenario could you see reverting the tests for the feature and not the feature?

**Comment by aaronemassey** [comment]
> In what scenario could you see reverting the tests for the feature and not the feature?

I see individual git commits as also a type of documentation for making changes to a particular system. In this PR's case, one commit could show to a reader an example of how to add something like `ZTEST_EXPECT_FAIL`. But if it's too much trouble don't worry about it I just think it's useful.

**Comment by yperess** [comment]
> > In what scenario could you see reverting the tests for the feature and not the feature?
> 
> I see individual git commits as also a type of documentation for making changes to a particular system. In this PR's case, one commit could show to a reader an example of how to add something like `ZTEST_EXPECT_FAIL`. But if it's too much trouble don't worry about it I just think it's useful.

I would prefer to keep them together in this case, the new API was added specifically to test the new behavior. It doesn't make sense to have them split up.

**Review by hakehuang — APPROVED** [review]
LGTM but need fixing the clang checking issue

**Comment by henrikbrixandersen** [comment]
> I see individual git commits as also a type of documentation for making changes to a particular system. In this PR's case, one commit could show to a reader an example of how to add something like `ZTEST_EXPECT_FAIL`. But if it's too much trouble don't worry about it I just think it's useful.

Separating changes into smaller commits also helps when reviewing a PR.

**Review by henrikbrixandersen — APPROVED** [review]
Thank you.

**Review by aaronemassey — APPROVED** [review]

**Review by andyross — APPROVED** [review]
> But... The very last statement from zmake tells twister if the whole binary passed/failed.

That too (the determination of whether the test as a whole succeeded or failed) is something that could be moved up into the outside rigging, though.  Really the "test framework" as a whole is always going to include both layers, ztest doesn't have to be an island.  But future design ideas aren't a -1, no reason not to merge this.

**Review by nashif — APPROVED** [review]

**Comment by henrikbrixandersen** [comment]
@andyross I think it's important to remember that ztest binaries can also be executed without being run via twister (during development, manual testing, other frameworks, ...). The test result needs to be available independently from twister.

**Comment by andyross** [comment]
I know, I do it it all the time. :)  I just think it's very reasonable to expect someone hand-running a test case to be able to read a "NOTE: this case expects failure" message manually, especially in the use case here which is limited to tests **of** the test framework itself.  Really I'm just reacting to the complexity of this patch vs. that level of simplicity.  And also banging the drum about how nice it would be if we had a unified "just run this test" library that west et. al. could use independently of twister.
